### PR TITLE
Created 'vud_widget' plugin type and 'plain' plugin

### PR DESCRIPTION
**Pull request by sudhanshug16**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+.idea
```
Comment (marvil07) [inline .gitignore]: This is unrelated, please only include related-to-topic changes on the topic branch.
Please notice that your suggestion for `.gitignore` is already part of upstream `8.x-1.x`, i.e. this is not needed if rebased.

File: vud_widget/src/Annotation/VoteUpDownWidget.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Defines a Vote Up/Down Field item annotation object.
+ *
+ * @see \Drupal\vud_widget\Plugin\VoteUpDownFieldManager
+ * @see plugin_api
+ *
+ * @Annotation
+ */
+class VoteUpDownWidget extends Plugin {
```
Comment (marvil07) [inline vud_widget/src/Annotation/VoteUpDownWidget.php:15]: This looks fine, but:
* again let's move it to vud module,
* please also include a couple more data members: widgetTemplate, and votesTemplate, see e.g. `widgets/plain/plain.inc`. This may change if we can replace `vud_pseudo_theming()` with real theme templating, but let's start with the same D7 approach.

File: vud_widget/src/Plugin/VoteUpDownWidget/plain.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ * Assigns ownership of a node to a user.
+ *
+ * @VotingApiWidget(
+ *   id = "plain",
+ *   label = @Translation("Plain rating widget")
+ *  )
+ */
+class Plain extends VoteUpDownWidgetBase
+{
+  use StringTranslationTrait;
```
Comment (marvil07) [inline vud_widget/src/Plugin/VoteUpDownWidget/plain.php:18]: This trait may be better on the base class instead.

File: vud_widget/src/Plugin/VoteUpDownWidgetInterface.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace Drupal\vud_widget\Plugin;
+
+use Drupal\Component\Plugin\PluginInspectionInterface;
+
+/**
+ * Defines an interface for Vote Up/Down Widget plugins.
+ */
+interface VoteUpDownWidgetInterface extends PluginInspectionInterface
```
Comment (marvil07) [inline vud_widget/src/Plugin/VoteUpDownWidgetInterface.php:10]: Please see all the calls to `ctools_plugin_get_function()` in D7, that may help to find the missing methods here.

File: vud_widget/src/Plugin/VoteUpDownWidgetBase.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+abstract class VoteUpDownWidgetBase extends PluginBase implements VoteUpDownWidgetInterface
+{
+
+  /**
+   * VoteUpDownWidgetBase constructor.
+   * @param array $configuration
+   * @param string $plugin_id
+   * @param mixed $plugin_definition
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition) {
```
Comment (marvil07) [inline vud_widget/src/Plugin/VoteUpDownWidgetBase.php:19]: Why is the constructor needed here? I see `PluginBase` class already has a constructor, and no customization is added.

File: vud_widget/src/Plugin/VoteUpDownWidgetInterface.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+{
+
+  /**
+   * @return mixed
```
Comment (marvil07) [inline vud_widget/src/Plugin/VoteUpDownWidgetInterface.php:14]: Please add documentation on docblocks. This is crucial for interfaces.

File: vud_widget/vud_widget.libraries.yml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+plain:
+  js:
+    js/vote.js : {}
+  css:
+    theme:
+      https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css: {}
```
Comment (marvil07) [inline vud_widget/vud_widget.libraries.yml:6]: Please do not include remote URLs.

**Review by marvil07 — CHANGES_REQUESTED** [review]
Thanks for starting this!
There are still several things to do.

I added detailed inline comments, but here some general notes:
* Why to move the widget definition from vud module to a new vud_widget module? In D7 the widgets were defined as ctools plugins inside the vud module, so I expect that in D8 the core plugin that will replace them should also be defined on vud module. Also this code is removing vud_field module, which is unrelated, please revert that.
* Related to the last one, please remove changes on vud_field.
* Please take the time to fill docblocks not only with types but with descriptions.

**Comment by marvil07** [comment]
Please also notice you may want to rebase this on top of current upstream `8.x-1.x`.

File: .gitignore (hunk 1 above)
Comment (sudhanshug16) [reply .gitignore]: Removed .gitignore

File: vud_widget/src/Annotation/VoteUpDownWidget.php (hunk 2 above)
Comment (sudhanshug16) [reply vud_widget/src/Annotation/VoteUpDownWidget.php:15]: Moved vud_widget to vud and added data types

File: vud_widget/src/Plugin/VoteUpDownWidget/plain.php (hunk 3 above)
Comment (sudhanshug16) [reply vud_widget/src/Plugin/VoteUpDownWidget/plain.php:18]: Moved trait to parent class

File: vud_widget/src/Plugin/VoteUpDownWidgetBase.php (hunk 5 above)
Comment (sudhanshug16) [reply vud_widget/src/Plugin/VoteUpDownWidgetBase.php:19]: Removed this constructor. I actually saw it in a post but never really got the logic. Seemed irrelevant to me too but kept it anyhow thinking about peer advice

File: vud_widget/src/Plugin/VoteUpDownWidgetInterface.php (hunk 6 above)
Comment (sudhanshug16) [reply vud_widget/src/Plugin/VoteUpDownWidgetInterface.php:14]: added docblocks

File: vud_widget/src/Plugin/VoteUpDownWidgetInterface.php (hunk 4 above)
Comment (sudhanshug16) [reply vud_widget/src/Plugin/VoteUpDownWidgetInterface.php:10]: A little confused. Elaborate please.

File: vud_widget/vud_widget.libraries.yml (hunk 7 above)
Comment (sudhanshug16) [reply vud_widget/vud_widget.libraries.yml:6]: Created a local font-awesome.min.css

**Review by sudhanshug16 — COMMENTED** [review]
Did the requested changes.Please review the commits :)

File: css/font-awesome.min.css [hunk 8]
```diff
@@ -0,0 +1,4 @@
+/*!
+ *  Font Awesome 4.7.0 by @davegandy - http://fontawesome.io - @fontawesome
+ *  License - http://fontawesome.io/license (Font: SIL OFL 1.1, CSS: MIT License)
```
Comment (marvil07) [inline css/font-awesome.min.css:3]: Please remove fontawesome. See general comment about this.

File: src/Plugin/VoteUpDownWidget/plain.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Assigns ownership of a node to a user.
+ *
+ * @VotingApiWidget(
+ *   id = "plain",
+ *   label = @Translation("Plain rating widget")
+ *  )
+ */
+class Plain extends VoteUpDownWidgetBase
```
Comment (marvil07) [inline src/Plugin/VoteUpDownWidget/plain.php:15]: File name should match the class name.

File: src/Plugin/VoteUpDownWidget/plain.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+{
```
Comment (marvil07) [inline src/Plugin/VoteUpDownWidget/plain.php:16]: Code standards error: place `{` in the same line of declaration. Please change this wherever is needed, including functions/methods.

File: src/Plugin/VoteUpDownWidgetBase.php [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+  use StringTranslationTrait;
+
+  /**
+   * VoteUpDownWidgetBase constructor.
+   * @param array $configuration
+   * @param string $plugin_id
+   * @param mixed $plugin_definition
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition) {
```
Comment (marvil07) [inline src/Plugin/VoteUpDownWidgetBase.php:22]: You mentioned this was removed, but it is still here.

File: src/Plugin/VoteUpDownWidgetInterface.php [hunk 12]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Drupal\vud\Plugin;
+
+use Drupal\Component\Plugin\PluginInspectionInterface;
+
+/**
+ * Defines an interface for Vote Up/Down Widget plugins.
+ */
+interface VoteUpDownWidgetInterface extends PluginInspectionInterface
```
Comment (marvil07) [inline src/Plugin/VoteUpDownWidgetInterface.php:10]:
```
# On D7
git grep ctools_plugin_get_function
vud.theme.inc:  if ($function = ctools_plugin_get_function($plugin, 'alter template variables')) {
vud.theme.inc:    if ($function = ctools_plugin_get_function($plugin, 'ajax render')) {
vud.theme.inc:      if ($function = ctools_plugin_get_function($plugin, 'ajax render')) {

```
i.e. a couple related methods missing on the interface.

File: vud_field/Annotation/VoteUpDownField.php [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Defines a Vote Up/Down Field item annotation object.
+ *
+ * @see \Drupal\vud_field\Plugin\VoteUpDownFieldManager
+ * @see plugin_api
+ *
+ * @Annotation
+ */
+class VoteUpDownField extends Plugin {
```
Comment (marvil07) [inline vud_field/Annotation/VoteUpDownField.php:15]: Please remove vud_field changes from this topic branch.

File: vud_widget.libraries.yml [hunk 14]
```diff
@@ -0,0 +1,11 @@
+plain:
```
Comment (marvil07) [inline vud_widget.libraries.yml:1]: Rename the file to vud, which is the related module name, otherwise it will not register.

File: vud_widget.services.yml [hunk 15]
```diff
@@ -0,0 +1,7 @@
+services:
```
Comment (marvil07) [inline vud_widget.services.yml:1]: idem, file renaming

File: vud_widget/css/plain.css [hunk 16]
```diff
@@ -0,0 +1,22 @@
+.vote-up-down{
```
Comment (marvil07) [inline vud_widget/css/plain.css:1]: I guess you forgot to remove the vud_widget directory?

**Review by marvil07 — CHANGES_REQUESTED** [review]
Thanks for the changes.

* Let's not change the widget to use fontawesome now, let's just use images as in D7. We may want to add an additional widget that works with that, but let's defer that till the end of a functional port.
* Please remove vud_field changes from this topic branch.
* Maybe you forgot to remove vud_widget/?
